Ical.Net/DataTypes/CalDateTime.cs

```diff
@@ -1,130 +1,185 @@
+    /// </summary>
+    /// <param name="kind">If <see langword="null"/>, <see cref="DateTimeKind.Unspecified"/> is used.</param>
+    /// <param name="tzId">The specified value will override value's <see cref="DateTime.Kind"/> property.
+    /// If a non-UTC time zone is specified, the underlying <see cref="DateTime.Kind"/> property will be <see cref="DateTimeKind.Local"/>.
+    /// If no time zone is specified, the <see cref="DateTime.Kind"/> property will be left untouched.
+    /// </param>
+    /// <param name="date"></param>
+    /// <param name="time"></param>
+    /// <param name="cal"></param>
+    public CalDateTime(DateOnly date, TimeOnly? time, DateTimeKind kind, string? tzId = null, Calendar? cal = null)
+    {
+        if (time.HasValue)
+            Initialize(new DateTime(date.Year, date.Month, date.Day, time.Value.Hour, time.Value.Minute, time.Value.Second, kind), true, tzId, cal);
+        else
+            Initialize(new DateTime(date.Year, date.Month, date.Day, 0, 0, 0, kind), false, tzId, cal);
     }
 
-    public CalDateTime(int year, int month, int day) : this(year, month, day, 0, 0, 0) { }
-    public CalDateTime(int year, int month, int day, string tzId) : this(year, month, day, 0, 0, 0, tzId) { }
-
+    /// <summary>
+    /// Creates a new instance of the <see cref="CalDateTime"/> class by parsing <paramref name="value"/>
+    /// using the <see cref="DateTimeSerializer"/>.
+    /// </summary>
+    /// <param name="value">An iCalendar-compatible date or date-time string.</param>
     public CalDateTime(string value)
     {
         var serializer = new DateTimeSerializer();
-        CopyFrom(serializer.Deserialize(new StringReader(value)) as ICopyable);
+        CopyFrom(serializer.Deserialize(new StringReader(value)) as ICopyable
+                 ?? throw new InvalidOperationException("Failure deserializing value"));
     }
 
-    private void Initialize(int year, int month, int day, int hour, int minute, int second, string tzId, Calendar cal)
+    private void Initialize(DateTime dateTime, bool hasTime, string? tzId, Calendar? cal)
     {
-        Initialize(CoerceDateTime(year, month, day, hour, minute, second, DateTimeKind.Local), tzId, cal);
-    }
+        DateTime initialValue;
 
-    private void Initialize(DateTime value, string tzId, Calendar cal)
-    {
-        if (!string.IsNullOrWhiteSpace(tzId) && !tzId.Equals("UTC", StringComparison.OrdinalIgnoreCase))
+        if ((tzId != null && !string.IsNullOrWhiteSpace(tzId) && !tzId.Equals("UTC", StringComparison.OrdinalIgnoreCase))
+            || (string.IsNullOrWhiteSpace(tzId) && dateTime.Kind == DateTimeKind.Local))
         {
             // Definitely local
-            value = DateTime.SpecifyKind(value, DateTimeKind.Local);
-            TzId = tzId;
+            _tzId = tzId;
+
+            initialValue = DateTime.SpecifyKind(dateTime, DateTimeKind.Local);
```

Comment:
Also legacy, but IMHO important: I think we should avoid `DateTimeKind.Local` wherever we can, because dealing with the system's local time really is a special case. If no tzid is specified, this would usually mean that the instance represents floating time (which btw cannot be converted to a different time zone). Floating time should rather be represented by `DateTimeKind.Unspecified` than `.Local`.

Reply:
Will be updated as proposed.

Reply:
Done